sips/sip-010/sip-010-fungible-token-standard.md

```diff
@@ -35,18 +35,24 @@ The fungible token trait, `ft-trait`, has a few methods:
 
 Transfer the fungible token from the sender of this transaction to the recipient. The `amount` is an unsigned integer. It is recommended that implementing contracts use the built-in `ft-transfer` Clarity method. If the sender does not have enough tokens to complete the transaction, the transaction should abort and return an `(err uint)`.
 
+This method must be defined with `define-public`, as it alters state, and should be externally callable.
+
 ### Name
 
 `(name () (response (string-ascii 32) uint))`
 
 Return a human-readable name for the contract, such as "CoolPoints", etc.
 
+This method should be defined as read-only, i.e. `define-read-only`.
```

Comment:
I should have been more specific, because in some cases, some functions must be defined as `define-public`, so a function callable via the `/v2/contracts/call-read/...` endpoint would be sufficient.

However, this may mean that the return value needs to be `(response)`, not just a primitive types.

The reason you can not always use `define-readonly` is that, and this is very unfortunate, `contrac-call?` is considered as a possible write operation, even when calling an other `define-read-only` function, preventing you from composing with other contracts in read-only functions.

Same for other read-only methods.

Reply:
> The reason you can not always use define-readonly is that, and this is very unfortunate, contrac-call? is considered as a possible write operation, even when calling an other define-read-only function, preventing you from composing with other contracts in read-only functions.

Damn, I didn't know that, and it is unfortunate.

> I should have been more specific, because in some cases, some functions must be defined as define-public, so a function callable via the /v2/contracts/call-read/... endpoint would be sufficient.

You're not saying that you _can_ call a `define-public` function via this endpoint, if it only does reads, can you? Let me try I suppose...

Reply:
There is an open issue to fix that: https://github.com/blockstack/stacks-blockchain/issues/1981

Reply:
In general, it should be possible to call a _write_ method via the `/v2/contracts/call-read/` endpoint.  But, the data won't be persisted to disk (the implementation keeps all pending writes in RAM until the transaction commits).

Reply:
Circling back to this, I think this comment thread is concerned about the interaction between read-only functions and traits, and less concerned with the trait definition itself.  I don't think the text can say that the method must be `define-read-only`, but it can recommend that the implementation be read-only.

We can write tooling that inspect an implementation's analysis data after it is deployed, and from there, infer whether or not the implementation is read-only.  I think this is something wallet authors and exchanges will do anyways, in order to help identify likely-buggy or likely-malicious token implementations.